Release tracking PR - v0.30.0

tcharding opened the pull request:
Add changelog notes and bump the version number to v0.30.0.

## TODO - pre-merge

- [x] Release `bitcoin_hashes` 0.12: https://github.com/rust-bitcoin/rust-bitcoin/pull/1694
- [x] Release secp 0.27: https://github.com/rust-bitcoin/rust-secp256k1/pull/588
  - https://github.com/rust-bitcoin/rust-secp256k1/pull/590
- [x] Update `secp256k1` dependency to use newly released v0.27: https://github.com/rust-bitcoin/rust-bitcoin/pull/1714 
- [x] Merge
  - ~https://github.com/rust-bitcoin/rust-bitcoin/pull/1696~
  - https://github.com/rust-bitcoin/rust-bitcoin/pull/1695
  -  https://github.com/rust-bitcoin/rust-bitcoin/pull/1111
- [x] If time permits merge these:
  - #1710 
  - #1705 
  - #1713 
- [x] Set the release date in changelog header
- [x] And merge these:
  - https://github.com/rust-bitcoin/rust-bitcoin/pull/1721
  - https://github.com/rust-bitcoin/rust-bitcoin/pull/1720
  - https://github.com/rust-bitcoin/rust-bitcoin/pull/1719
  - https://github.com/rust-bitcoin/rust-bitcoin/pull/1717  

## TODO  - post release
- [ ] Release the blogpost: https://github.com/rust-bitcoin/www.rust-bitcoin.org/pull/2
   - ~Set the date in the blog post to match the date 0.30 is released~

tcharding commented:
Force pushed some more changelog notes re updating dependencies we control.

Kixunil commented:
Note for Optech: should be published after this is merged and the crate is out. Should be soon.

apoelstra commented:
@tcharding would you be willing to open a PR to do the actual update?

apoelstra commented:
Oh, never mind, this *is* the PR -- except we just need a PR to update the rust-secp dep. I can do that.

apoelstra commented:
Done: https://github.com/rust-bitcoin/rust-bitcoin/pull/1714

tcharding commented:
Are we happy with the taproot API improvements under the "API improvements" header and not in their own section? (Mentioning because we just added a separate taproot section to the blog post.)

tcharding commented:
@apoelstra I've set the date to Sunday the 19th March so if we release in the next day or two the "set date" checkbox in PR description can be considered done.

Kixunil commented on bitcoin/CHANGELOG.md:
```diff
@@ -1,11 +1,63 @@
+
+- We moved a lot of modules around, specifically we move basically everything out of `util`.
+
+- API improvements:
```
Unsized `Script` is not mentioned here but I think it should since it's quite large and disruptive

tcharding replied:
Added here and also in the "New types" section.

Kixunil commented:
> Are we happy with the taproot API improvements under the "API improvements" header and not in their own section? (Mentioning because we just added a separate taproot section to the blog post.)

If you mean in changelog then yes, they don't have to be the same.

tcharding commented on bitcoin/CHANGELOG.md:
```diff
@@ -1,11 +1,63 @@
-  contains a `LockTime` but is `Ord`, we have manually sorted the locktimes based on
-  their consensus encoding. This ordering is somewhat arbitrary -- there is no total
-  ordering on locktimes since they may be measured in either blocks or seconds.
+This release is big, to help users upgrade we wrote a blog post, please see https://rust-bitcoin.org/
```
I did this vague link on purpose to introduce folks to the new website.

Kixunil replied:
Having a separate "Hey, we have a website!" would make more sense then.

tcharding replied:
Sure thing, will add.

tcharding commented:
Force push is just the mentions of `ScriptBuf`.

tcharding commented:
Rebased on master and added initial link to the new website. Used direct link to blogpost. We have a chicken and egg problem with the blog post and changelog now, we (I) want to add links to online docs to the blog post which implies code released (changelog notes released) which link to the blogpost.

Kixunil commented:
IMO start by releasing blog first because less people will accidentally stumble upon it.

tcharding commented:
Aight, I rekon we are good to pull the trigger then - either @apoelstra to do it (blogpost and crate release) or I can be online to do it with you Andrew at normal time. I'm getting pretty excited about this release FTR.

tcharding commented:
Bother, too excited it would seem. We still need to merge #1714 and then rebase this one.

tcharding commented:
Updated the release date in changelog to 21st

apoelstra reviewed (dismissed):
ACK 99d816948adf03633ae3aac1be64fed45628416c

tcharding commented:
Needs an ack from you @Kixunil!

tcharding commented:
Perhaps give @sanket1729 a chance to ack as well?

apoelstra commented:
If we update this (which we should, unless we get two more ACKs on this commit quickly), we may also want to mention https://github.com/rust-bitcoin/rust-bitcoin/pull/1715 in the changelog (which I just merged)

junderw commented:
I just noticed something missing from all the float work. See #1720

It's a small change, so I would hope it can get in. Sorry for the last minute. I was reviewing the format PR and noticed Header had a passthrough method.

Kixunil commented:
Such a shame, 21st March is my lucky day, would've been be really cool to release today. :D

I'm OK with fixing it up but since #1715 merged I'd also like to make `Denomination` `non_exhaustive`.

Kixunil reviewed (dismissed):
ACK 31f0f832ad5f13852cf96f3f4fb94d9256822866

@tcharding had a good point about timezones, it's worth releasing with imperfect changelog.

sanket1729 reviewed (approved):
ACK 31f0f832ad5f13852cf96f3f4fb94d9256822866

tcharding commented on bitcoin/CHANGELOG.md:
```diff
@@ -1,11 +1,68 @@
+     ordering on locktimes since they may be measured in either blocks or seconds.
+
+- Performance improvements:
+  - [Add `FromHexStr`](https://github.com/rust-bitcoin/rust-bitcoin/pull/1400) for parsing strings with and without `0x` prefix
```
I moved the `FromHexStr` line up to "API improvements" section and change performance section to:

- Performance improvements:
  - [Remove needless allocation from BIP-158 encoding](https://github.com/rust-bitcoin/rust-bitcoin/pull/1146)
  - [Implement fast hex encoding](https://github.com/rust-bitcoin/rust-bitcoin/pull/1268) (usage added in a [later PR](https://github.com/rust-bitcoin/rust-bitcoin/pull/1476))

tcharding commented on bitcoin/CHANGELOG.md:
```diff
@@ -1,11 +1,68 @@
+  - [Try to fix up sighash export mess](https://github.com/rust-bitcoin/rust-bitcoin/pull/1277)
+
+- New types:
+  - `ScriptBuf` ([Implement unsized `Script`](https://github.com/rust-bitcoin/rust-bitcoin/pull/1155).
```
Changed to 

  - Renamed `Script` to `ScriptBuf` and added a new unsized type [`Script`](https://github.com/rust-bitcoin/rust-bitcoin/pull/1155)

tcharding commented on bitcoin/CHANGELOG.md:
```diff
@@ -1,11 +1,68 @@
+  - [New `Witness` features](https://github.com/rust-bitcoin/rust-bitcoin/pull/1323)
+  - [`Witness` API improvements](https://github.com/rust-bitcoin/rust-bitcoin/pull/1380)
+  - [Add newtype `ScriptBuf`](https://github.com/rust-bitcoin/rust-bitcoin/pull/1155) (unsized
+    `Script`) and use `&Script` for function argument when borrowing a script.
```
Removed and is now the line posted in thread above.

sanket1729 reviewed (approved):
reACK ffee8ad81ba16d3ccc15b82e898adc3a67931f76

tcharding commented:
reack please @sanket1729 and @Kixunil if you are still about - I rekon we can still get this out today!

Kixunil reviewed (approved):
ACK ffee8ad81ba16d3ccc15b82e898adc3a67931f76

apoelstra reviewed (approved):
ACK ffee8ad81ba16d3ccc15b82e898adc3a67931f76

tcharding commented:
Wo ho ho, getting close ...

apoelstra commented:
Just started merge process. should be less than 30 minutes.

tcharding commented:
Can you merge the blog post PR now then? I can deploy if you like (if I do it then I can fix it if its not working :)

tcharding commented:
In order for the blog post to be live when we release I'm going to deploy it now.

apoelstra commented:
@tcharding yes, the remaining thing to do is merge this PR :). But I have to run my CI job on the merge commit before I can do that.

apoelstra commented:
codex32 https://github.com/BlockstreamResearch/codex32/tree/master/reference/rust-codex32 has a `Fe` type which is the same as `u5` in this library.

It's not obvious that either lib should depend on the other though so actually sharing a type doesn't seem clear.

sanket1729 commented:
Let's go!!!

apoelstra commented:
@sanket1729 sorry, we have one more thing to fix :) the `readme` key in Cargo.toml needs us to add a `../` because we moved the workspace.

Tobin will PR, then if we can quickl ACK (I won't bother doing any tests) we should be able to push in a few minutes.

tcharding commented:
#1723

apoelstra commented:
And we're live!

sanket1729 commented:
Is the website also up? I would like tweet about the release.

tcharding commented:
I've borked the website, trying to fix now ...

sanket1729 commented:
Wohoo. Seems to be up now https://rust-bitcoin.github.io/blog/release-0.30.0/

apoelstra commented:
Looks like a 404 to me

tcharding commented:
yeah I have no idea why. I've hard reset to same commit as was up before and settings in https://github.com/rust-bitcoin/rust-bitcoin.github.io/settings/pages look correct

tcharding commented:
Looks like its stuck in githubs github pages build system

tcharding commented:
Its through their build system now but 404 still

tcharding commented:
It was up the whole time, cache on my machine was returning the 404

tcharding commented:
face palm, still 404 - this is working: https://rust-bitcoin.github.io/index.html

tcharding commented:
We are live yo!
